src/test/java/org/mule/service/http/impl/service/client/HttpClientStreamingTestCase.java

```diff
@@ -0,0 +1,272 @@
+import org.mule.tck.probe.Probe;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import ru.yandex.qatools.allure.annotations.Description;
+import ru.yandex.qatools.allure.annotations.Features;
+import ru.yandex.qatools.allure.annotations.Stories;
+
+@Features(HTTP_SERVICE)
+@Stories(STREAMING)
+@Description("Validates HTTP client behaviour against a streaming server.")
+public class HttpClientStreamingTestCase extends AbstractMuleTestCase {
+
+  @Rule
+  public DynamicPort serverPort = new DynamicPort("serverPort");
+
+  private static final int RESPONSE_SIZE = 14 * 1024;
+  private static final int WAIT_TIMEOUT = 5000;
+  private static final int RESPONSE_TIMEOUT = 3000;
+  private static final int TIMEOUT_MILLIS = 1000;
+  private static final int POLL_DELAY_MILLIS = 200;
+
+  private static Latch latch;
+
+  private HttpServiceImplementation service = new HttpServiceImplementation(new SimpleUnitTestSupportSchedulerService());
+  private HttpClientConfiguration.Builder clientBuilder = new HttpClientConfiguration.Builder().setName("streaming-test");
+  private PollingProber pollingProber = new PollingProber(TIMEOUT_MILLIS, POLL_DELAY_MILLIS);
+  private HttpServer server;
+
+  @Before
+  public void setUp() throws Exception {
+    latch = new Latch();
+    service.start();
+    server = service.getServerFactory().create(new HttpServerConfiguration.Builder()
+        .setHost("localhost")
+        .setPort(serverPort.getNumber())
+        .setName("streaming-test")
+        .build());
+    server.start();
+    server.addRequestHandler("/",
+                             (requestContext, responseCallback) -> responseCallback.responseReady(setUpHttpResponse(),
+                                                                                                  new IgnoreResponseStatusCallback()));
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (server != null) {
+      server.stop();
+      server.dispose();
+    }
+    service.stop();
+  }
+
+  @Test
+  @Description("Uses a streaming HTTP client to send a non blocking request which will finish before the stream is released.")
+  public void nonBlockingStreaming() throws Exception {
+    HttpClient client = service.getClientFactory().create(clientBuilder.build());
+    client.start();
+    final Reference<HttpResponse> responseReference = new Reference<>();
+    try {
+      client.send(getRequest(), RESPONSE_TIMEOUT, true, null, new ResponseHandler() {
+
+        @Override
+        public void onCompletion(HttpResponse response) {
+          responseReference.set(response);
+        }
+
+        @Override
+        public void onFailure(Exception exception) {
+          latch.release();
+          fail("Response should have been received despite latch being set.");
```

Comment:
Will this fail actually fail test, because it'll be executed by a different thread.